### PR TITLE
MAYA-115097 - I'd like to define my Maya Reference in a variant on cr…

### PR DESCRIPTION
**Pull request by seando-adsk**
#### MAYA-115097 - I'd like to define my Maya Reference in a variant on creation

* Added python/mel string translation utilities to core (lib folder). Register any "lib" strings from the MayaUsd plugin.
* Added all the "Usd Options" in the "Add Maya Reference" dialog. Only the "Define in Variant" section is functional at this time.
* Updated mayaUsdAddMayaReference option vars to work in all versions of Maya we support.
* Moved "sanitize name" code from rename command to helper with python binding and unit test.

### CODE REVIEW

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel [hunk 1]
```diff
@@ -1,25 +1,80 @@
...
+}
 
-                frameLayout -defineTemplate optionsTemplate 
-                        -collapsable true
-                        -collapse false
-                        -labelVisible true 
-                        -labelIndent 5 
-                        -marginWidth 5 
-                        -marginHeight 5;
+proc createRowLayoutforMayaReference(string $parent, string $rowLayoutName)
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel:40]: Helper proc to create a row layout

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel [hunk 2]
```diff
@@ -251,61 +364,246 @@ proc setOptionVars()
...
-    //   "reject", reserve old one and reject new.
-    //   The default option is "new", to ensure not to destroy the file data
-    optionVar -init $forceFactorySettings -category $category
-        -sv namespaceClashNameMethod "new"
-        -sv fileWorkingNamespaceName ""
-        ;
-    
+    int $mjv = `about -majorVersion`;
+    if ($mjv <= 2022)
+    {
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel:369]: Handle older versions of Maya. The "optionVar --init" flag was added after Maya 2022.

**Review by seando-adsk — COMMENTED** [review]
@ppt-adsk If you have any questions about the code we can meet over zoom to discuss them.

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel [hunk 3]
```diff
@@ -251,61 +364,246 @@ proc setOptionVars()
...
+{
+    global string $gUsdMayaReferenceUfePath;
+    global int $gSingleWidgetWidthIndex;
+
+    // If the caller didn't give us the "Group" checkbox status, get it now.
+    if ($groupChecked == -1) {
+        $groupChecked = `checkBoxGrp -q -v1 usdMayaReferenceGroupGrp`;
+    }
+
+    // If the caller didn't give us the variane set name size, get it now.
```
Comment (ppt-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel:522]: "variane"

File: test/lib/mayaUsd/fileio/testAddMayaReference.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        tempMayaFile = os.path.join(tempfile.gettempdir(), 'simpleSphere.ma')
+        cmds.file(rename=tempMayaFile)
+        cmds.file(save=True, force=True, type='mayaAscii')
+        return tempMayaFile
+
+    def testDefineInVariant(self):
+        '''Test the "Define in Variant" options for Add Maya Reference.'''
+
+        # First we need a pure Maya scene to reference in.
+        mayaSceneStr = self.createSimpleMayaScene()
```
Comment (ppt-adsk) [inline test/lib/mayaUsd/fileio/testAddMayaReference.py:64]: Shouldn't we clean this up once the test is done?

File: test/lib/mayaUsd/fileio/testAddMayaReference.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+        # Make sure the prim has the variant set and variant.
+        self.assertTrue(primA.HasVariantSets())
+        vset = primA.GetVariantSet(kDefaultVariantSetName)
+        self.assertTrue(vset.IsValid())
+        self.assertEqual(vset.GetName(), kDefaultVariantSetName)
+        self.assertTrue(vset.GetVariantNames())
+        self.assertTrue(vset.HasAuthoredVariant(kDefaultVariantName))
+        self.assertEqual(vset.GetVariantSelection(), kDefaultVariantName)
+
```
Comment (ppt-adsk) [inline test/lib/mayaUsd/fileio/testAddMayaReference.py:95]: Would be good to actually check that the Maya reference prim is inside the variant, and that it has the expected metadata.

**Review by ppt-adsk — COMMENTED** [review]
Great work!  Only a couple of minor comments.

File: test/lib/mayaUsd/fileio/testAddMayaReference.py (hunk 4 above)
Comment (seando-adsk) [reply test/lib/mayaUsd/fileio/testAddMayaReference.py:64]: No need really. It is created in a folder (specific to this test) in the "temporary" testing folder for all tests.

**Review by ppt-adsk — APPROVED** [review]

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py [hunk 6]
```diff
@@ -14,14 +41,39 @@ def createMayaReferencePrim(ufePathStr, mayaReferencePath, mayaNamespace):
     primPath = Sdf.AssetPath(parentPath+'/MayaReference')
 
     stage = mayaUsd.ufe.getStage(ufePathStr)
-    prim = stage.DefinePrim(primPath.path, 'MayaReference')
+
+    variantSetName = Tf.MakeValidIdentifier(variantSetName)
+    variantName = Tf.MakeValidIdentifier(variantName)
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.py:46]: Use the Pixar Usd function.

**Review by ppt-adsk — APPROVED** [review]

**Review by seando-adsk — COMMENTED** [review]
No need for our own 'sanitizeName' method - I found functions in Usd that do exactly that. And these are better since internally its what they use to test and sanitize the names.

**Review by ppt-adsk — DISMISSED** [review]

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel [hunk 7]
```diff
@@ -455,9 +753,48 @@ global proc addMayaReferenceToUsdUiCb(string $parent, string $selectedFile)
...
+            string $varSetName = `optionMenu -q -value usdMayaReferenceVariantSetNameOptionMenu`;
+            if ($varSetName == $createNew) {
+                optionVar -sv usdMayaReferenceVariantSetNameOption $varSetNameText;
+            } else {
+                optionVar -sv usdMayaReferenceVariantSetNameOption $varSetName;
+            }
+        } else {
+            optionVar -sv usdMayaReferenceVariantSetNameOption $varSetNameText;
+        }
+        optionVar -sv usdMayaReferenceVariantSetNameText $varSetNameText;
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel:780]: Sorry another change here regarding the optionVar values after a create new (spec update from Natalia).

File: lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel [hunk 8]
```diff
@@ -629,12 +993,159 @@ global proc string addMayaReferenceToUsd(string $ufePath)
...
+    if ($value == "") {
+        string $defValue = getMayaUsdLibString("kMayaRefDefaultVariantSetName");
+        textField -edit -text $defValue usdMayaReferenceVariantSetNameTextField;
+    } else {
+        // Make sure the name user entered doesn't contain any invalid characters.
+        string $validatedSetName = sanitizeName($value);
+        if ($value != $validatedSetName) {
+            textField -edit -text $validatedSetName usdMayaReferenceVariantSetNameTextField;
+        }
+    }
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel:1135]: And here - sanitize the variant set and variant names after user enters them, so the optionVar will have the correct value as well.

File: test/lib/mayaUsd/fileio/testAddMayaReference.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        # Create another prim to test sanitizing variant name.
+        primB = stage.DefinePrim('/B', 'Xform')
+        primPathStr = proxyShapePathStr + ',/B'
+        primB = mayaUsd.ufe.ufePathToPrim(primPathStr)
+
+        kBadVariantSetName = 'No Spaces or Special#Chars'
+        kGoodVariantSetName = Tf.MakeValidIdentifier(kBadVariantSetName)
+        kBadVariantName = '3no start digits'
+        kGoodVariantName = Tf.MakeValidIdentifier(kBadVariantName)
```
Comment (seando-adsk) [inline test/lib/mayaUsd/fileio/testAddMayaReference.py:121]: Fixed test to use same sanitize from usd.

**Review by seando-adsk — COMMENTED** [review]
Sorry a couple more changes.

**Review by ppt-adsk — APPROVED** [review]

**Pull request merged**